addons/mail/models/mail_thread.py

```diff
@@ -3675,25 +3675,38 @@ def _notify_thread_by_web_push(self, message, recipients_data, msg_vals=False, *
 
         msg_vals = dict(msg_vals or {})
         partner_ids = self._extract_partner_ids_for_notifications(message, msg_vals, recipients_data)
-        if not partner_ids:
+        push_parameters = self._get_web_push_parameters(partner_ids)
+        if not push_parameters:
             return
+        payload = self._truncate_payload(self._notify_by_web_push_prepare_payload(message, msg_vals=msg_vals))
+        self._push_web_notification(payload, push_parameters)
 
-        partner_devices_sudo = self.env['mail.push.device'].sudo()
-        devices = partner_devices_sudo.search([
-            ('partner_id', 'in', partner_ids)
-        ])
-        if not devices:
+    def _get_web_push_parameters(self, partner_ids: list) -> tuple | None:
+        """
+        :param partner_ids: IDs of the res.partners
+        :returns: the `mail.push.device` records, the vapid private key and the vapid public key,
+        None if any is missing.
+        """
+        if not partner_ids:
             return
```

Comment:
Keep a single return type, it eases using the method.
```suggestion
            return self.env['mail.push.device'].sudo(), False, False
```